week4/assignment/auth-practice/src/hooks/useJoin.ts

```diff
@@ -0,0 +1,149 @@
+import { postJoin } from '@apis/member';
+import { formatPhoneNumber, validatePassword } from '@utils/utils';
+import React, { useRef, useState } from 'react';
+import { useNavigate } from 'react-router-dom';
+
+const useJoin = () => {
+  const navigate = useNavigate();
+
+  // input focus를 위한 ref
+  const idRef = useRef<HTMLInputElement>(null);
+  const passwordRef = useRef<HTMLInputElement>(null);
+  const nicknameRef = useRef<HTMLInputElement>(null);
+  const phoneNumberRef = useRef<HTMLInputElement>(null);
+
+  // id, password, nickname, phoneNumber를 관리하는 state
+  const [id, setId] = useState('');
+  const [password, setPassword] = useState('');
+  const [nickname, setNickname] = useState('');
+  const [phoneNumber, setPhoneNumber] = useState('');
+
+  // id, password, nickname, phoneNumber 에러 여부를 관리하는 state
+  const [idError, setIdError] = useState(false);
+  const [passwordError, setPasswordError] = useState(false);
+  const [nicknameError, setNicknameError] = useState(false);
+  const [phoneNumberError, setPhoneNumberError] = useState(false);
+
+  // id input이 변경될 때 호출되는 함수
+  const handleChangeId = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setId(e.target.value);
+  };
+  // password input이 변경될 때 호출되는 함수
+  const handleChangePassword = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setPassword(e.target.value);
+  };
+  // nickname input이 변경될 때 호출되는 함수
+  const handleChangeNickname = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setNickname(e.target.value);
+  };
+  // phoneNumber input이 변경될 때 호출되는 함수
+  const handleChangePhoneNumber = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setPhoneNumber(formatPhoneNumber(e.target.value));
+  };
```

Comment:
p5: 저는 항상 이렇게 input이 많을 때, name과 value를 사용해서 하나의 함수로 만드는 것 같아요! 
```suggestion
  const handleChangeInput = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = e.target;

    if (name === 'id') {
      setId(value);
    } else if (name === 'password') {
      setPassword(value);
    } else if (name === 'nickname') {
      setNickname(value);
    } else if (name === 'phone') {
      setPhoneNumber(formatPhoneNumber(value));
    }
  };
```
근데 사실 이렇게 적고, 태승님 코드를 보니까 태승님은 하나의 함수가 한 가지 역할만 하도록 정말 잘 작성하신 것 같더라구요...
근데 이게 그렇게 되면 너무 함수를 많이 만들어야할 것 같은데, 적당한 면에서는 같은 역할을 하는 애들을 묶어도 될 것 같아요!
이 부분은 나중에 같이 얘기해보면 너무 좋을 것 같아요 ㅎ

Reply:
저도 사실 이부분 name으로 처리하신 도영님 코드 보고 되게 좋다고 생각했었습니다. 좋은 방법 공유 감사하고 저도 고민해보겠습니다!
그리고 리뷰에 코드 제안 하는게 있네요.. 이런기능도 첨알았습니다 ㅎㅎ 감사합니다

Reply:
저도 도영님 코드 보고 효율성이 좋다고 생각했었는데 이렇게 refector할수 있다는 점배워갑니다.

Reply:
이야 좋은 인사이트 공유네요 두가지 방법 다 심도있게 고민해보게 되는 아주 좋은 코드리뷰에요